# 2.1 The Python Interpreter

Python is an interpreted language. The Python interpreter runs a program by executing one statement at a time. The standard interactive Python interpreter can be invoked on the command line with the python command:

The >>> you see is the prompt after which you’ll type code expressions. To exit the Python interpreter, you can either type exit() or press Ctrl-D (works on Linux and macOS only).

Running Python programs is as simple as calling python with a .py file as its first argument. Suppose we had created hello_world.py with these contents:

You can run it by executing the following command (the hello_world.py file must be in your current working terminal directory):

## IPython shell

While some Python programmers execute all of their Python code in this way, those doing data analysis or scientific computing make use of IPython, an enhanced Python interpreter, or Jupyter notebooks, web-based code notebooks originally created within the IPython project. 

When you use the <b>%run</b> command, IPython executes the code in the specified file in the same process, enabling you to explore the results interactively when it’s done



# Running the Jupyter Notebook

One of the major components of the Jupyter project is the notebook, a type of interactive document for code, text (including Markdown), data visualizations, and other output. The Jupyter notebook interacts with kernels, which are implementations of the Jupyter interactive computing protocol specific to different programming languages. The Python Jupyter kernel uses the IPython system for its underlying behavior.

## Tab Completion


On the surface, the IPython shell looks like a cosmetically different version of the standard terminal Python interpreter (invoked with python). One of the major improvements over the standard Python shell is tab completion, found in many IDEs or other interactive computing analysis environments. While entering expressions in the shell, pressing the Tab key will search the namespace for any variables (objects, functions, etc.) matching the characters you have typed so far and show the results in a convenient drop-down menu:

In [6]:
an_apple = 27

an_example = 42

In [ ]:
an

In this example, note that IPython displayed both of the two variables I defined, as well as the built-in function any. Also, you can also complete methods and attributes on any object after typing a period:

In [8]:
b = [1, 2, 3]

In [ ]:
b.

The same is true for modules:

In [10]:
import datetime

In [ ]:
datetime.

Note that IPython by default hides methods and attributes starting with underscores, such as magic methods and internal “private” methods and attributes, in order to avoid cluttering the display (and confusing novice users!). These, too, can be tab-completed, but you must first type an underscore to see them. If you prefer to always see such methods in tab completion, you can change this setting in the IPython configuration. See the IPython documentation to find out how to do this.

Tab completion works in many contexts outside of searching the interactive namespace and completing object or module attributes. When typing anything that looks like a file path (even in a Python string), pressing the <b>Tab</b> key will complete anything on your computer’s filesystem matching what you’ve typed.

Combined with the <b>%run</b> command (see “The %run Command”), this functionality can save you many keystrokes.

Another area where tab completion saves time is in the completion of function keyword arguments (including the = sign!). See Figure 2-4.

![](pda3_0204.png)

We’ll have a closer look at functions in a little bit.

## Introspection

Using a question mark (?) before or after a variable will display some general information about the object:

In [13]:
b = [1, 2, 3]

In [15]:
b?

Type:        list
String form: [1, 2, 3]
Length:      3
Docstring:  
Built-in mutable sequence.

If no argument is given, the constructor creates a new empty list.
The argument must be an iterable if specified.

In [25]:
print?

Signature: print(*args, sep=' ', end='\n', file=None, flush=False)
Docstring:
Prints the values to a stream, or to sys.stdout by default.

sep
  string inserted between values, default a space.
end
  string appended after the last value, default a newline.
file
  a file-like object (stream); defaults to the current sys.stdout.
flush
  whether to forcibly flush the stream.
Type:      builtin_function_or_method

This is referred to as object introspection. If the object is a function or instance method, the docstring, if defined, will also be shown. Suppose we’d written the following function (which you can reproduce in IPython or Jupyter):

In [17]:
def add_numbers(a, b):
    """
    Add two numbers together

    Returns
    -------
    the_sum : type of arguments
    """
    return a + b

Then using <b>?</b> shows us the docstring:

In [20]:
add_numbers?

Signature: add_numbers(a, b)
Docstring:
Add two numbers together

Returns
-------
the_sum : type of arguments
File:      /var/folders/r3/gmt08cnn7lzdwnng743vjxmw0000gp/T/ipykernel_7336/1411870314.py
Type:      function

<b>?</b> has a final usage, which is for searching the IPython namespace in a manner similar to the standard Unix or Windows command line. A number of characters combined with the wildcard (*) will show all names matching the wildcard expression. For example, we could get a list of all functions in the top-level NumPy namespace containing load:

In [23]:
import numpy as np

In [25]:
np.*load*?

np.__loader__
np.load
np.loadtxt

# 2.3 Python Language Basics

In this section, I will give you an overview of essential Python programming concepts and language mechanics. In the next chapter, I will go into more detail about Python data structures, functions, and other built-in tools.

## Language Semantics

The Python language design is distinguished by its emphasis on readability, simplicity, and explicitness. Some people go so far as to liken it to “executable pseudocode.”

### Indentation, not braces

Python uses whitespace (tabs or spaces) to structure code instead of using braces as in many other languages like R, C++, Java, and Perl. Consider a for loop from a sorting algorithm:

A <b>colon</b> denotes the start of an indented code block after which all of the code must be indented by the same amount until the end of the block.

<blockquote>Note
    
I strongly recommend using four spaces as your default indentation and replacing tabs with four spaces. Many text editors have a setting that will replace tab stops with spaces automatically (do this!). IPython and Jupyter notebooks will automatically insert four spaces on new lines following a colon and replace tabs by four spaces.</blockquote>

As you can see by now, Python statements also do not need to be terminated by semicolons. <b>Semicolons</b> can be used, however, to separate multiple statements on a single line:

In [55]:
a = 5; b = 6; c = 7

Putting multiple statements on one line is generally discouraged in Python as it can make code less readable.

## Everything is an object

An important characteristic of the Python language is the consistency of its object model. Every number, string, data structure, function, class, module, and so on exists in the Python interpreter in its own “box,” which is referred to as a Python object. Each object has an associated type (e.g., integer, string, or function) and internal data. In practice this makes the language very flexible, as even functions can be treated like any other object.

## Comments

Any text preceded by the hash mark (pound sign) <b>#</b> is ignored by the Python interpreter. This is often used to add comments to code. At times you may also want to exclude certain blocks of code without deleting them. One solution is to comment out the code:

In [64]:
results = []
file_handle = []
for line in file_handle:
    # keep the empty lines for now
    # if len(line) == 0:
    #   continue
    results.append(line.replace("foo", "bar"))

Comments can also occur after a line of executed code. While some programmers prefer comments to be placed in the line preceding a particular line of code, this can be useful at times:

In [67]:
print("Reached this line")  # Simple status report

Reached this line


## Function and object method calls

You call functions using parentheses and passing zero or more arguments, optionally assigning the returned value to a variable:

Almost every object in Python has attached functions, known as methods, that have access to the object’s internal contents. You can call them using the following syntax:

Functions can take both positional and keyword arguments:

We will look at this in more detail later.

## Variables and argument passing

When assigning a variable in Python, you are creating a reference to the object shown on the right-hand side of the equals sign. In practical terms, consider a list of integers:

In [29]:
a = [1, 2, 3]

Suppose we assign <b>a</b> to a new variable <b>b</b>:

In [32]:
b=a
b

[1, 2, 3]

In some languages, the assignment of <b>b</b> will cause the data [1, 2, 3] to be copied. In Python, <b>a</b> and <b>b</b> actually now refer to the same object, the original list [1, 2, 3] (see Figure 2-5 for a mock-up). You can prove this to yourself by appending an element to <b>b</b> and then examining <b>a</b>:

In [124]:
b.append(4)
a

[1, 2, 3, 4]

![](pda3_0205.png)

Understanding the semantics of references in Python, and when, how, and why data is copied, is especially critical when you are working with larger datasets in Python.

<blockquote>Assignment is also referred to as binding, as we are binding a name to an object. Variable names that have been assigned may occasionally be referred to as bound variables.</blockquote>

When you pass objects as arguments to a function, new local variables are created referencing the original objects without any copying. If you bind a new object to a variable inside a function, that will not overwrite a variable of the same name in the “scope” outside of the function (the “parent scope”). It is therefore possible to alter the internals of a mutable argument. Suppose we had the following function:

In [128]:
def append_element(some_list, element):
    some_list.append(element)

Then we have:

In [96]:
data = [1, 2, 3]
append_element(data, 4)
data

[1, 2, 3, 4]

### Dynamic references, strong types

Variables in Python have no inherent type associated with them; a variable can refer to a different type of object simply by doing an assignment. There is no problem with the following:

In [100]:
a = 5
type(a)

int

In [102]:
a = "foo"
type(a)

str

Variables are names for objects within a particular namespace; the type information is stored in the object itself. Some observers might hastily conclude that Python is not a “typed language.” This is not true; consider this example:

In [105]:
"5" + 5

TypeError: can only concatenate str (not "int") to str

In some languages, the string '5' might get implicitly converted (or cast) to an integer, thus yielding 10. In other languages the integer 5 might be cast to a string, yielding the concatenated string '55'. In Python, such implicit casts are not allowed. In this regard we say that Python is a strongly typed language, which means that every object has a specific type (or class), and implicit conversions will occur only in certain permitted circumstances, such as:

In [108]:
a = 4.5

b = 2

# String formatting, to be visited later
print(f"a is {type(a)}, b is {type(b)}")

a is <class 'float'>, b is <class 'int'>


In [110]:
a / b

2.25

Here, even though b is an integer, it is implicitly converted to a float for the division operation.

Knowing the type of an object is important, and it’s useful to be able to write functions that can handle many different kinds of input. You can check that an object is an instance of a particular type using the isinstance function:

In [113]:
a = 5
isinstance(a, int)

True

isinstance can accept a tuple of types if you want to check that an object’s type is among those present in the tuple:

In [35]:
a = 5; b = 4.5

In [43]:
isinstance(a, (int, float))

True

In [47]:
isinstance(b, (int, str))

False

## Attributes and methods

Objects in Python typically have both <b>attributes</b> (other Python objects stored “inside” the object) and <b>methods</b> (functions associated with an object that can have access to the object’s internal data). Both of them are accessed via the syntax <b>obj.attribute_name</b>:

In [50]:
a = "foo"

Try pressing \<tab\> after an object

In [58]:
a.

SyntaxError: invalid syntax (3905452595.py, line 1)

Attributes and methods can also be accessed by name via the getattr function:

In [60]:
getattr(a, "split")

<function str.split(sep=None, maxsplit=-1)>

While we will not extensively use the functions getattr and related functions hasattr and setattr in this book, they can be used very effectively to write generic, reusable code.

## Duck typing

Often you may not care about the type of an object but rather only whether it has certain methods or behavior. This is sometimes called duck typing, after the saying “If it walks like a duck and quacks like a duck, then it’s a duck.” For example, you can verify that an object is iterable if it implements the iterator protocol. For many objects, this means it has an __iter__ “magic method,” though an alternative and better way to check is to try using the iter function:

In [76]:
def isiterable(obj):
     try:
         iter(obj)
         return True
     except TypeError: # not iterable
         return False

This function would return True for strings as well as most Python collection types:

In [79]:
isiterable("a string")

True

In [81]:
isiterable([1, 2, 3])

True

In [83]:
isiterable(5)

False

## Modules

In Python, a module is simply a file with the .py extension containing Python code. Suppose we had the following module:

If we wanted to access the variables and functions defined in some_module.py, from another file in the same directory we could do:

In [67]:
import some_module
#result = some_module.f(5)
pi = some_module.PI
pi

3.14159

Or alternately:

In [70]:
from some_module import g, PI

In [72]:
result = g(5, PI)
print(result)

8.14159


By using the <b>as</b> keyword, you can give imports different variable names:

In [106]:
import some_module as sm
from some_module import PI as pi, g as gf

In [110]:
r1 = sm.f(pi)
print(r1)

5.14159


In [112]:
r2 = gf(6, pi)
print(r2)

9.14159


## Binary operators and comparisons

Most of the binary math operations and comparisons use familiar mathematical syntax used in other programming languages:

In [170]:
5 - 7

-2

In [172]:
12 + 21.5

33.5

In [174]:
5 <= 2

False

See Table 2-1 for all of the available binary operators.

|Operation|Description|
|---|---|
|a + b|Add a and b|
|a - b|Subtract b from a|
|a * b|Multiply a by b|
|a / b|Divide a by b|
|a // b|Floor-divide a by b, dropping any fractional remainder|
|a ** b|Raise a to the b power|
|a & b|	True if both a and b are True; for integers, take the bitwise AND|
|a \| b|	True if either a or b is True; for integers, take the bitwise OR|
|a ^ b|	For Booleans, True if a or b is True, but not both; for integers, take the bitwise EXCLUSIVE-OR|
|a == b|True if a equals b|
|a != b|	True if a is not equal to b|
|a < b, a <= b|	True if a is less than (less than or equal to) b|
|a > b, a >= b|	True if a is greater than (greater than or equal to) b|
|a is b|	True if a and b reference the same Python object|
|a is not b|	True if a and b reference different Python objects|

To check if two variables <b>refer</b> to the same object, use the <b>is</b> keyword. 

Use <b>is not</b> to check that two objects are not the same:

In [82]:
a = [1, 2, 3]

b = a

c = list(a)

In [85]:
a is b

True

In [90]:
a is c

False

In [87]:
a is not c

True

The <b>list</b> function always creates a new Python list (i.e., a copy), so <b>c</b> and <b>a</b> do not refer to the same object. 

However, comparing <b>a</b> and <b>c</b> with the <b>==</b> operator shows that their contents are equal :

In [95]:
a == c

True

A common use of <b>is</b> and <b>is not</b> is to check if a variable is <b>None</b>, since there is only one instance of <b>None</b>.

This can be useful for scanning a document for missing values

In [195]:
a = None
a is None

True

## Mutable and immutable objects

Many objects in Python, such as lists, dictionaries, NumPy arrays, and most user-defined types (classes), are mutable. This means that the object or values that they contain can be modified:

In [99]:
a_list = ["foo", 2, [4, 5]]
a_list[2] = (3, 4)
a_list

['foo', 2, (3, 4)]

Others, like strings and tuples, are immutable, which means their internal data cannot be changed:

In [202]:
a_tuple = (3, 5, (4, 5))

a_tuple[1] = "four"

TypeError: 'tuple' object does not support item assignment

Remember that just because you can mutate an object does not mean that you always should. Such actions are known as side effects. For example, when writing a function, any side effects should be explicitly communicated to the user in the function’s documentation or comments. If possible, I recommend trying to avoid side effects and favor immutability, even though there may be mutable objects involved.

## Scalar Types

Python has a small set of built-in types for handling numerical data, strings, Boolean (True or False) values, and dates and time. These “single value” types are sometimes called scalar types, and we refer to them in this book as scalars . See Table 2-2 for a list of the main scalar types. Date and time handling will be discussed separately, as these are provided by the datetime module in the standard library.

|Type|Description|
|---|---|
|None|The Python “null” value (only one instance of the None object exists)|
|str|String type; holds Unicode strings|
|bytes|Raw binary data|
|float|Double-precision floating-point number (note there is no separate double type)|
|bool|A Boolean True or False value|
|int|Arbitrary precision integer|

## Numeric types

The primary Python types for numbers are int and float. An int can store arbitrarily large numbers:

In [210]:
ival = 17239871

ival ** 6

26254519291092456596965462913230729701102721

Floating-point numbers are represented with the Python float type. Under the hood, each one is a double-precision value. They can also be expressed with scientific notation:

In [213]:
fval = 7.243

fval2 = 6.78e-5

Integer division not resulting in a whole number will always yield a floating-point number:

In [216]:
3 / 2

1.5

To get C-style integer division (which drops the fractional part if the result is not a whole number), use the floor division operator //:

In [219]:
3 // 2

1

## Strings

Many people use Python for its built-in string handling capabilities. You can write string literals using either single quotes <b>'</b> or double quotes <b>"</b> (double quotes are generally favored):

In [111]:
a = 'one way of writing a string'
b = "another way"

The Python string type is <b>str</b>.

For multiline strings with line breaks, you can use triple quotes, either <b>'''</b> or <b>"""</b>:

In [114]:
c = """
This is a longer string that
spans multiple lines
"""

It may surprise you that this string c actually contains four lines of text; the line breaks after """ and after lines are included in the string. We can count the new line characters with the count method on c:

In [117]:
c.count("\n")

3

Python strings are immutable; you cannot modify a string:

In [120]:
a = "this is a string"

a[10] = "f"

TypeError: 'str' object does not support item assignment

To interpret this error message, read from the bottom up. We tried to replace the character at position 10 with the letter "f", but this is not allowed for string objects. If we need to modify a string, we have to use a function or method that creates a new string, such as the string replace method:

In [122]:
b = a.replace("string", "longer string")
b

'this is a longer string'

Afer this operation, the variable <b>a</b> is unmodified:

In [125]:
a

'this is a string'

Many Python objects can be converted to a string using the <b>str</b> function:

In [128]:
a = 5.6

s = str(a)

print(s)

5.6


Strings are a sequence of Unicode characters and therefore can be treated like other sequences, such as lists and tuples:

In [134]:
s = "python"

list(s)

['p', 'y', 't', 'h', 'o', 'n']

In [136]:
s[:3]

'pyt'

The syntax <b>s[:3]</b> is called <b>slicing</b> and is implemented for many kinds of Python sequences. This will be explained in more detail later on, as it is used extensively in the book.

The backslash character <b>\\</b> is an escape character, meaning that it is used to specify special characters like newline <b>\\n</b> or Unicode characters. To write a string literal with backslashes, you need to escape them:

In [142]:
s = "12\\34"

print(s)

12\34


If you have a string with a lot of backslashes and no special characters, you might find this a bit annoying. Fortunately you can preface the leading quote of the string with <b>r</b>, which means that the characters should be interpreted as is:

In [145]:
s = r"this\has\no\special\characters"

s

'this\\has\\no\\special\\characters'

The <b>r</b> stands for raw.

Adding two strings together concatenates them and produces a new string:

In [148]:
a = "this is the first half "

b = "and this is the second half"

a + b

'this is the first half and this is the second half'

String templating or formatting is another important topic. The number of ways to do so has expanded with the advent of Python 3, and here I will briefly describe the mechanics of one of the main interfaces. 

String objects have a format method that can be used to substitute formatted arguments into the string, producing a new string:

In [172]:
template = "{0:.2f} {1:s} are worth US${2:d}"

In this string:

<b>{0:.2f}</b> means to format the <b>first</b> argument as a <b>floating-point</b> number with two decimal places.

<b>{1:s}</b> means to format the <b>second</b> argument as a <b>string</b>.

<b>{2:d}</b> means to format the <b>third</b> argument as an exact <b>integer</b>.

To substitute arguments for these format parameters, we pass a sequence of arguments to the format method:

In [175]:
 template.format(88.46, "Argentine Pesos", 1)

'88.46 Argentine Pesos are worth US$1'

Python 3.6 introduced a new feature called f-strings (short for formatted string literals) which can make creating formatted strings even more convenient. To create an f-string, write the character f immediately preceding a string literal. Within the string, enclose Python expressions in curly braces to substitute the value of the expression into the formatted string:

In [182]:
amount = 10

rate = 88.46

currency = "Pesos"

result = f"{amount} {currency} is worth US${amount / rate}"

Format specifiers can be added after each expression using the same syntax as with the string templates above:

In [184]:
f"{amount} {currency} is worth US${amount / rate:.2f}"

'10 Pesos is worth US$0.11'

String formatting is a deep topic; there are multiple methods and numerous options and tweaks available to control how values are formatted in the resulting string. To learn more, consult the official Python documentation.

## Bytes and Unicode

In modern Python (i.e., Python 3.0 and up), Unicode has become the first-class string type to enable more consistent handling of ASCII and non-ASCII text. In older versions of Python, strings were all bytes without any explicit Unicode encoding. You could convert to Unicode assuming you knew the character encoding. Here is an example Unicode string with non-ASCII characters:

In [273]:
val = "español"

val

'español'

We can convert this Unicode string to its UTF-8 bytes representation using the encode method:

In [276]:
val_utf8 = val.encode("utf-8")

In [278]:
val_utf8

b'espa\xc3\xb1ol'

In [280]:
type(val_utf8)

bytes

Assuming you know the Unicode encoding of a bytes object, you can go back using the decode method:

In [283]:
val_utf8.decode("utf-8")

'español'

While it is now preferable to use UTF-8 for any encoding, for historical reasons you may encounter data in any number of different encodings:

In [286]:
val.encode("latin1")

b'espa\xf1ol'

In [288]:
val.encode("utf-16")

b'\xff\xfee\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

In [290]:
val.encode("utf-16le")

b'e\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

It is most common to encounter bytes objects in the context of working with files, where implicitly decoding all data to Unicode strings may not be desired.

In [293]:
True and True

True

In [295]:
False or True

True

When converted to numbers, False becomes 0 and True becomes 1:

In [298]:
int(False)

0

In [300]:
int(True)

1

The keyword <b>not</b> flips a Boolean value from True to False or vice versa:

In [303]:
a = True
b = False

In [305]:
not a

False

In [307]:
not b

True

## Type casting

The str, bool, int, and float types are also functions that can be used to cast values to those types:

In [38]:
s = "3.14159"

fval = float(s)
fval

3.14159

In [313]:
type(fval)

float

In [315]:
int(fval)

3

In [317]:
bool(fval)

True

In [319]:
bool(0)

False

Note that most nonzero values when cast to <b>bool</b> become <b>True</b>.

## None

<b>None</b> is the Python null value type:

In [324]:
a = None

In [326]:
a is None

True

In [328]:
b = 5

In [330]:
b is not None

True

<b>None</b> is also a common default value for function arguments:

In [333]:
def add_and_maybe_multiply(a, b, c=None):
    result = a + b

    if c is not None:
        result = result * c

    return result

## Dates and times

The built-in Python datetime module provides datetime, date, and time types. The datetime type combines the information stored in date and time and is the most commonly used:

In [337]:
from datetime import datetime, date, time

dt = datetime(2011, 10, 29, 20, 30, 21)

In [339]:
dt.day

29

In [341]:
dt.minute

30

Given a datetime instance, you can extract the equivalent date and time objects by calling methods on the datetime of the same name:

In [344]:
dt.date()

datetime.date(2011, 10, 29)

In [346]:
dt.time()

datetime.time(20, 30, 21)

The strftime method formats a datetime as a string:

In [349]:
dt.strftime("%Y-%m-%d %H:%M")

'2011-10-29 20:30'

Strings can be converted (parsed) into datetime objects with the strptime function:

In [352]:
datetime.strptime("20091031", "%Y%m%d")

datetime.datetime(2009, 10, 31, 0, 0)

See Table 11-2 for a full list of format specifications.

When you are aggregating or otherwise grouping time series data, it will occasionally be useful to replace time fields of a series of datetimes—for example, replacing the minute and second fields with zero:

In [355]:
dt_hour = dt.replace(minute=0, second=0)

dt_hour

datetime.datetime(2011, 10, 29, 20, 0)

Since datetime.datetime is an immutable type, methods like these always produce new objects. So in the previous example, dt is not modified by replace:

In [358]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

The difference of two datetime objects produces a datetime.timedelta type:

In [361]:
dt2 = datetime(2011, 11, 15, 22, 30)

delta = dt2 - dt

In [363]:
delta

datetime.timedelta(days=17, seconds=7179)

In [365]:
type(delta)

datetime.timedelta

The output timedelta(17, 7179) indicates that the timedelta encodes an offset of 17 days and 7,179 seconds.

Adding a timedelta to a datetime produces a new shifted datetime:

In [368]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [370]:
dt + delta

datetime.datetime(2011, 11, 15, 22, 30)

## Control Flow

Python has several built-in keywords for conditional logic, loops, and other standard control flow concepts found in other programming languages.

## Conditionals
### if, elif, and else


The if statement is one of the most well-known control flow statement types. It checks a condition that, if True, evaluates the code in the block that follows:

In [376]:
x = -5
if x < 0:
    print("It's negative")

It's negative


An if statement can be optionally followed by one or more elif blocks and a catchall else block if all of the conditions are False:

In [379]:
if x < 0:
    print("It's negative")
elif x == 0:
    print("Equal to zero")
elif 0 < x < 5:
    print("Positive but smaller than 5")
else:
    print("Positive and larger than or equal to 5")

It's negative


If any of the conditions are True, no further elif or else blocks will be reached. With a compound condition using and or or, conditions are evaluated left to right and will short-circuit:

In [382]:
a = 5; b = 7

c = 8; d = 4

if a < b or c > d:
    print("Made it")

Made it


In this example, the comparison c > d never gets evaluated because the first comparison was True.

It is also possible to chain comparisons:

In [385]:
4 > 3 > 2 > 1

True

### for loops

<b>for</b> loops are for iterating over a collection (like a list or tuple) or an iterater. The standard syntax for a <b>for</b> loop is:

You can advance a for loop to the next iteration, skipping the remainder of the block, using the continue keyword. Consider this code, which sums up integers in a list and skips None values:

In [193]:
sequence = [1, 2, None, 4, None, 5]
total = 0
for value in sequence:
    if value is None:
        continue
    total += value

In [195]:
total

12

A for loop can be exited altogether with the break keyword. This code sums elements of the list until a 5 is reached:

In [197]:
sequence = [1, 2, 0, 4, 6, 5, 2, 1]
total_until_5 = 0
for value in sequence:
    if value == 5:
        break
    total_until_5 += value

The break keyword only terminates the innermost for loop; any outer for loops will continue to run:

In [398]:
for i in range(4):
    for j in range(4):
        if j > i:
            break
        print((i, j))

(0, 0)
(1, 0)
(1, 1)
(2, 0)
(2, 1)
(2, 2)
(3, 0)
(3, 1)
(3, 2)
(3, 3)


As we will see in more detail, if the elements in the collection or iterator are sequences (tuples or lists, say), they can be conveniently unpacked into variables in the for loop statement:

In [401]:
for a, b, c in iterator:
    # do something

SyntaxError: incomplete input (3249532412.py, line 2)

### while loops

A while loop specifies a condition and a block of code that is to be executed until the condition evaluates to False or the loop is explicitly ended with break:

In [405]:
x = 256
total = 0
while x > 0:
    if total > 500:
        break
    total += x
    x = x // 2

### pass

pass is the “no-op” (or “do nothing”) statement in Python. It can be used in blocks where no action is to be taken (or as a placeholder for code not yet implemented); it is required only because Python uses whitespace to delimit blocks:

In [409]:
if x < 0:
    print("negative!")
elif x == 0:
    # TODO: put something smart here
    pass
else:
    print("positive!")

positive!


### range

The range function generates a sequence of evenly spaced integers:

In [413]:
range(10)

range(0, 10)

In [415]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

A start, end, and step (which may be negative) can be given:

In [418]:
list(range(0, 20, 2))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

In [420]:
list(range(5, 0, -1))

[5, 4, 3, 2, 1]

As you can see, range produces integers up to but not including the endpoint. A common use of range is for iterating through sequences by index:

In [423]:
seq = [1, 2, 3, 4]

for i in range(len(seq)):
    print(f"element {i}: {seq[i]}")

element 0: 1
element 1: 2
element 2: 3
element 3: 4


While you can use functions like list to store all the integers generated by range in some other data structure, often the default iterator form will be what you want. This snippet sums all numbers from 0 to 99,999 that are multiples of 3 or 5:

In [426]:
total = 0

for i in range(100_000):
    # % is the modulo operator
    if i % 3 == 0 or i % 5 == 0:
        total += i

print(total)

2333316668


While the range generated can be arbitrarily large, the memory use at any given time may be very small.

# 2.4 Conclusion

This chapter provided a brief introduction to some basic Python language concepts and the IPython and Jupyter programming environments. In the next chapter, I will discuss many built-in data types, functions, and input-output utilities that will be used continuously throughout the rest of the book.